File: app/src/main/java/com/example/diffutilrv/view/EmployeeRecyclerViewAdapter.java
package com.example.diffutilrv.view;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.diffutilrv.R;
import com.example.diffutilrv.model.Employee;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

public class EmployeeRecyclerViewAdapter extends ListAdapter<Employee, EmployeeRecyclerViewAdapter.ViewHolder> {

    @Inject
    protected EmployeeRecyclerViewAdapter(DiffUtil.ItemCallback<Employee> diffCallback) {
        super(diffCallback);
    }


    @NotNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        final View view = inflater.inflate(R.layout.list_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull EmployeeRecyclerViewAdapter.ViewHolder holder, int position) {
        final Employee employee = getItem(position);
        holder.name.setText(employee.getName());
        holder.role.setText(employee.getRole());
    }


    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView role;
        private final TextView name;

        public ViewHolder(View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.employee_name);
            role = itemView.findViewById(R.id.employee_role);
        }
    }
}

File: app/src/main/java/com/example/diffutilrv/EmployeeRecyclerViewAdapter.kt
package com.example.diffutilrv

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView

class EmployeeRecyclerViewAdapter(employeeList: List<Employee>?) :
    RecyclerView.Adapter<EmployeeRecyclerViewAdapter.ViewHolder>() {
    private val mEmployees: MutableList<Employee> = ArrayList()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.list_item, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val (_, name, role) = mEmployees[position]
        holder.name.text = name
        holder.role.text = role
    }

    fun updateEmployeeListItems(employees: List<Employee>?) {
        val diffCallback = EmployeeDiffCallback(mEmployees, employees)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        mEmployees.clear()
        mEmployees.addAll(employees!!)
        diffResult.dispatchUpdatesTo(this)
    }

    override fun getItemCount(): Int {
        return mEmployees.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val role: TextView
        val name: TextView

        init {
            name = itemView.findViewById<View>(R.id.employee_name) as TextView
            role = itemView.findViewById<View>(R.id.employee_role) as TextView
        }
    }

    init {
        mEmployees.addAll(employeeList!!)
    }
}